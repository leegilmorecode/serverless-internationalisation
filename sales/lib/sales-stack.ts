import * as apigw from "aws-cdk-lib/aws-apigateway";
import * as nodeLambda from "aws-cdk-lib/aws-lambda-nodejs";

import { Stack, StackProps } from "aws-cdk-lib";

import { Construct } from "constructs";

interface SalesProps extends StackProps {
  apiProps: apigw.RestApiProps;
  getSalesLambdaProps: nodeLambda.NodejsFunctionProps;
  createSaleLambdaProps: nodeLambda.NodejsFunctionProps;
}

export class SalesStack extends Stack {
  constructor(scope: Construct, id: string, props?: SalesProps) {
    super(scope, id, props);

    // lambda resources
    const getSalesHandler: nodeLambda.NodejsFunction =
      new nodeLambda.NodejsFunction(this, "GetSales", {
        ...props?.getSalesLambdaProps,
      });

    const createSaleHandler: nodeLambda.NodejsFunction =
      new nodeLambda.NodejsFunction(this, "CreateSale", {
        ...props?.createSaleLambdaProps,
      });

    // api gateway resource
    const salesApi: apigw.RestApi = new apigw.RestApi(this, "SalesAPI", {
      ...props?.apiProps,
    });

    const sales: apigw.Resource = salesApi.root.addResource("sales");
    sales.addMethod(
      "GET",
      new apigw.LambdaIntegration(getSalesHandler, {
        proxy: true,
        allowTestInvoke: true,
      })
    );

    sales.addMethod(
      "POST",
      new apigw.LambdaIntegration(createSaleHandler, {
        proxy: true,
        allowTestInvoke: true,
      })
    );
  }
}
