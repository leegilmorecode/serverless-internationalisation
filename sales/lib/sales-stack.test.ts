import * as cdk from "aws-cdk-lib";

import { SalesStack } from "./sales-stack";
import { Template } from "aws-cdk-lib/assertions";

describe("sales-stack", () => {
  const OLD_ENV = process.env;

  beforeEach(() => {
    jest.resetModules();
    process.env = { ...OLD_ENV };
  });

  afterAll(() => {
    process.env = OLD_ENV;
  });

  describe("Global", () => {
    it("should generate the correct global lambdas", async () => {
      process.env.LOCALE = "global";
      const { dynamicImport } = require("@shared/dynamic-imports");

      const { apiProps } = await dynamicImport("@lib/global/api-gateway");
      const { getSalesLambdaProps } = await dynamicImport(
        "@lib/global/lambdas"
      );

      const app = new cdk.App();
      const stack = new SalesStack(app, "GlobalTestStack", {
        apiProps,
        getSalesLambdaProps,
      });

      const template = Template.fromStack(stack);

      expect(
        template.findResources("AWS::Lambda::Function", {
          Properties: {
            FunctionName: "getSalesLambdaProps",
          },
        })
      ).toMatchSnapshot();
    });

    it("should generate the correct global api gateway", async () => {
      process.env.LOCALE = "global";
      const { dynamicImport } = require("@shared/dynamic-imports");

      const { apiProps } = await dynamicImport("@lib/global/api-gateway");
      const { getSalesLambdaProps } = await dynamicImport(
        "@lib/global/lambdas"
      );

      const app = new cdk.App();
      const stack = new SalesStack(app, "GlobalTestStack", {
        apiProps,
        getSalesLambdaProps,
      });

      const template = Template.fromStack(stack);

      expect(
        template.findResources("AWS::ApiGateway::RestApi")
      ).toMatchSnapshot();
    });
  });

  describe("UK", () => {
    it("should generate the correct UK lambdas", async () => {
      process.env.LOCALE = "UK";
      const { dynamicImport } = require("@shared/dynamic-imports");

      const { apiProps } = await dynamicImport("@lib/global/api-gateway");
      const { getSalesLambdaProps } = await dynamicImport(
        "@lib/global/lambdas"
      );

      const app = new cdk.App();
      const stack = new SalesStack(app, "UKTestStack", {
        apiProps,
        getSalesLambdaProps,
      });

      const template = Template.fromStack(stack);

      expect(
        template.findResources("AWS::Lambda::Function", {
          Properties: {
            FunctionName: "getSalesLambdaProps",
          },
        })
      ).toMatchSnapshot();
    });

    it("should generate the correct UK api gateway", async () => {
      process.env.LOCALE = "UK";
      const { dynamicImport } = require("@shared/dynamic-imports");

      const { apiProps } = await dynamicImport("@lib/global/api-gateway");
      const { getSalesLambdaProps } = await dynamicImport(
        "@lib/global/lambdas"
      );

      const app = new cdk.App();
      const stack = new SalesStack(app, "GlobalTestStack", {
        apiProps,
        getSalesLambdaProps,
      });

      const template = Template.fromStack(stack);

      expect(
        template.findResources("AWS::ApiGateway::RestApi")
      ).toMatchSnapshot();
    });
  });
});
