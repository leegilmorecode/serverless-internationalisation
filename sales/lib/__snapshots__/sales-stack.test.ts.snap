// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sales-stack Global should generate the correct global api gateway 1`] = `
Object {
  "SalesAPI40AE10D2": Object {
    "Properties": Object {
      "Description": "Global version of the Sales API",
      "Name": "sales-api",
    },
    "Type": "AWS::ApiGateway::RestApi",
  },
}
`;

exports[`sales-stack Global should generate the correct global lambdas 1`] = `
Object {
  "GetSalesD8890AF0": Object {
    "DependsOn": Array [
      "GetSalesServiceRole776C3A71",
    ],
    "Properties": Object {
      "Code": Object {
        "S3Bucket": Object {
          "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
        },
        "S3Key": "0e11a6c893f6fb4e262b2146d4ddef861a0d50933b9eadea21fdbc4f4f04dbf4.zip",
      },
      "Description": "Global version of the getSalesLambdaProps handler",
      "Environment": Object {
        "Variables": Object {
          "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          "LOCALE": "global",
        },
      },
      "FunctionName": "getSalesLambdaProps",
      "Handler": "index.handler",
      "MemorySize": 512,
      "Role": Object {
        "Fn::GetAtt": Array [
          "GetSalesServiceRole776C3A71",
          "Arn",
        ],
      },
      "Runtime": "nodejs14.x",
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;

exports[`sales-stack UK should generate the correct UK api gateway 1`] = `
Object {
  "SalesAPI40AE10D2": Object {
    "Properties": Object {
      "Description": "UK version of the Sales API",
      "Name": "sales-api",
    },
    "Type": "AWS::ApiGateway::RestApi",
  },
}
`;

exports[`sales-stack UK should generate the correct UK lambdas 1`] = `
Object {
  "GetSalesD8890AF0": Object {
    "DependsOn": Array [
      "GetSalesServiceRole776C3A71",
    ],
    "Properties": Object {
      "Code": Object {
        "S3Bucket": Object {
          "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
        },
        "S3Key": "0e11a6c893f6fb4e262b2146d4ddef861a0d50933b9eadea21fdbc4f4f04dbf4.zip",
      },
      "Description": "UK version of the getSalesLambdaProps handler",
      "Environment": Object {
        "Variables": Object {
          "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          "LOCALE": "uk",
        },
      },
      "FunctionName": "getSalesLambdaProps",
      "Handler": "index.handler",
      "MemorySize": 1024,
      "Role": Object {
        "Fn::GetAtt": Array [
          "GetSalesServiceRole776C3A71",
          "Arn",
        ],
      },
      "Runtime": "nodejs14.x",
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;
